src/flownet/data/perforation_strategy.py

```diff
@@ -0,0 +1,278 @@
+
+    Returns:
+        Dataframe with 1 or more connections per will depending on the historic straddles / plugs.
+
+    """
+    df = multiple(df)
+
+    df_w_groups = pd.DataFrame(
+        [], columns=["WELL_NAME", "X", "Y", "Z", "DATE", "OPEN", "GROUPID"]
+    )
+    df_groups = pd.DataFrame([], columns=["X", "Y", "Z", "GROUPID"])
+    groupid = 0
+
+    # Step 1
+    for well_name in df["WELL_NAME"].unique():
+        df_well = df.loc[df["WELL_NAME"] == well_name][
+            ["X", "Y", "Z", "WELL_NAME", "OPEN", "DATE"]
+        ]
+        df_well_piv = df_well.pivot_table("OPEN", ["X", "Y", "Z", "WELL_NAME"], "DATE")
+        df_well_piv.fillna(method="ffill", axis=1, inplace=True)
+        df_well_piv.fillna(False, inplace=True)
+        df_well_piv = df_well_piv.apply(lambda x: hash(tuple(x)), axis=1)
+
+        for group in df_well_piv.unique():
+            df_group = (
+                df_well_piv.loc[df_well_piv == group]
+                .index.to_frame()
+                .reset_index(drop=True)[["X", "Y", "Z"]]
+            )
+            df_group["GROUPID"] = groupid
+            groupid += 1
+            df_groups = df_groups.append(df_group)
+
+        df_w_groups = df_w_groups.append(
+            df_well.merge(df_groups, how="left", on=["X", "Y", "Z"])
+        )
+
+    # Step 2
+    for groupid in df_w_groups["GROUPID"].unique():
+        df_group = df_w_groups.loc[df_w_groups["GROUPID"] == groupid]
+
+        xmin, ymin, zmin = df_group[["X", "Y", "Z"]].min()
+        xmax, ymax, zmax = df_group[["X", "Y", "Z"]].max()
+
+        df_foreign = df_w_groups.loc[
+            (
+                ((df_w_groups["X"] > xmin) & (df_w_groups["X"] < xmax))
+                | ((df_w_groups["Y"] > ymin) & (df_w_groups["Y"] < ymax))
+                | ((df_w_groups["Z"] > zmin) & (df_w_groups["Z"] < zmax))
+            )
+            & (df_w_groups["GROUPID"] != groupid)
+        ]
+
+        # Step 3
+        if df_foreign.shape[0]:
+            xmin_foreign, ymin_foreign, zmin_foreign = df_foreign[["X", "Y", "Z"]].min()
+            xmax_foreign, ymax_foreign, zmax_foreign = df_foreign[["X", "Y", "Z"]].max()
+
+            df_w_groups.loc[
+                (df_w_groups["GROUPID"] == groupid)
+                & (
+                    (df_w_groups["X"] < xmin_foreign)
+                    | (df_w_groups["Y"] < ymin_foreign)
+                    | (df_w_groups["Z"] < zmin_foreign)
+                ),
+                ["GROUPID"],
+            ] = (
+                df_w_groups["GROUPID"].max() + 1
+            )
+
+            df_w_groups.loc[
+                (df_w_groups["GROUPID"] == groupid)
+                & (
+                    (df_w_groups["X"] > xmax_foreign)
+                    | (df_w_groups["Y"] > ymax_foreign)
+                    | (df_w_groups["Z"] > zmax_foreign)
+                ),
+                ["GROUPID"],
+            ] = (
+                df_w_groups["GROUPID"].max() + 1
+            )
+
```

Comment:
For a long deviated well with all connections opening/closing at the same times the bounding box defined by (xmin, xmax, ymin, ymax, zmin, zmax) can get quite large. Can this potentially make us split connections we don't want to split? Are there better options for a bounding box? Some kind of cylindrical area along the well? I guess we can add it as some kind of issue/improvement suggestion?

Reply:
I agree, the bounding box solution is quite poor. I have indeed also thought about fitting some spline and trying o fine out the order like that. However, then there are also cases with multi laterals that are even harder to identify with the definition used in the simulator.

If you have a better solution that is still -relatively- straightforward to implement I'd happily include it in the PR as I'm not quite satisfied with the bounding box. Suggestions are welcome 🍪 

(We would be rid of this problem immediately when we go to the real data though.)

Reply:
Issue created #223